cdap-gateway/src/test/java/co/cask/cdap/gateway/handlers/StreamHandlerTestV3.java

```diff
@@ -17,17 +17,163 @@
+    for (int i = 0; i < 2; ++i) {
+      sendEvent(streamId2, Integer.toString(i));
+    }
+
+    // Check metrics to verify that the metric for events processed is specific to each stream
+    checkEventsProcessed(streamId1, 10, 10);
+    checkEventsProcessed(streamId2, 2, 10);
+  }
+
+
+  private void createStream(Id.Stream streamId) throws Exception {
+    URL url = createURL(streamId.getNamespaceId(), "streams/" + streamId.getName());
+    HttpRequest request = HttpRequest.put(url).build();
+    HttpResponse response = HttpRequests.execute(request);
+    Assert.assertEquals(200, response.getResponseCode());
+  }
+
+  private void sendEvent(Id.Stream streamId, String body) throws Exception {
+    URL url = createURL(streamId.getNamespaceId(), "streams/" + streamId.getName());
+    HttpRequest request = HttpRequest.post(url).withBody(body).build();
+    HttpResponse response = HttpRequests.execute(request);
+    Assert.assertEquals(200, response.getResponseCode());
+  }
+
+  private List<String> fetchEvents(Id.Stream streamId) throws Exception {
+    URL url = createURL(streamId.getNamespaceId(), "streams/" + streamId.getName() + "/events");
+    HttpRequest request = HttpRequest.get(url).build();
+    HttpResponse response = HttpRequests.execute(request);
+    Assert.assertEquals(200, response.getResponseCode());
+
+
+    List<String> events = Lists.newArrayList();
+    JsonArray jsonArray = new JsonParser().parse(response.getResponseBodyAsString()).getAsJsonArray();
+    for (JsonElement jsonElement : jsonArray) {
+      events.add(jsonElement.getAsJsonObject().get("body").getAsString());
+    }
+    return events;
+  }
+  
+  private void checkEventsProcessed(Id.Stream streamId, int expectedCount, int retries) throws Exception {
+    for (int i = 0; i < retries; i++) {
+      int numProcessed = getNumProcessed(streamId);
+      if (numProcessed == expectedCount) {
+        return;
+      }
+      TimeUnit.SECONDS.sleep(1);
+    }
+    Assert.assertEquals(expectedCount, getNumProcessed(streamId));
+  }
+
+
+  private int getNumProcessed(Id.Stream streamId) throws Exception {
+    String path = String.format("/v3/metrics/query?metric=system.collect.events&context=ns.%s.str.%s",
+                                streamId.getNamespaceId(), streamId.getName());
+    System.out.println(path);
+    HttpRequest request = HttpRequest.post(getEndPoint(path).toURL()).build();
+    HttpResponse response = HttpRequests.execute(request);
+    Assert.assertEquals(200, response.getResponseCode());
+    System.out.println(response.getResponseBodyAsString());
+    return getNumEventsFromResponse(response.getResponseBodyAsString());
+  }
+
+  private int getNumEventsFromResponse(String response) {
+    // response format:
+    // [{"measureName":"system.collect.events","tagValues":{},"timeValues":[{"timestamp":0,"value":12}]}]
+
+    JsonElement jsonResponse = new JsonParser().parse(response);
```

Comment:
i would suggest using gson here,
```new Gson().fromJson(EntityUtils.toString(response.getEntity()), MetricQueryResult.class);```
and then read the MetriQueryResult for the value.